src/meta/src/hummock/manager/mod.rs

```diff
@@ -1348,18 +1340,35 @@ impl HummockManager {
 
     fn is_compact_task_expired(
         compact_task: &CompactTask,
-        branched_ssts: &BTreeMap<HummockSstableObjectId, BranchedSstInfo>,
+        hummock_version: &HummockVersion,
     ) -> bool {
-        for input_level in compact_task.get_input_ssts() {
-            for table_info in input_level.get_table_infos() {
-                if let Some(mp) = branched_ssts.get(&table_info.object_id) {
-                    if mp
-                        .get(&compact_task.compaction_group_id)
-                        .map_or(true, |sst_id| *sst_id != table_info.sst_id)
-                    {
-                        return true;
+        if let Some(group) = hummock_version
+            .levels
+            .get(&compact_task.compaction_group_id)
+        {
+            for input_level in compact_task.get_input_ssts() {
+                let input_level: &InputLevel = input_level;
+                let mut sst_ids: HashSet<_> = input_level
+                    .table_infos
+                    .iter()
+                    .map(|sst| sst.sst_id)
+                    .collect();
+                fn filter_ssts(levels: &Level, sst_ids: &mut HashSet<u64>) {
+                    for sst in &levels.table_infos {
+                        sst_ids.remove(&sst.sst_id);
                     }
                 }
+                if input_level.level_idx == 0 {
+                    for level in &group.get_level0().sub_levels {
+                        filter_ssts(level, &mut sst_ids);
+                    }
+                } else {
+                    filter_ssts(group.get_level(input_level.level_idx as _), &mut sst_ids);
+                }
+                if !sst_ids.is_empty() {
+                    warn!(stale_sst_id = ?sst_ids, ?compact_task, "compact task expired");
```

Comment:
I don't think this line should ever be reached if there's no bug, i.e. this is_compact_task_expired is no longer necessary.
@Li0k @Little-Wallace

Reply:
And it's kind of compute-intensive because it need to iterate through all SSTs in corresponding level.

Reply:
I think the cost is acceptable. That's how we currently do it when `apply_version_delta`.